spring-subscription-callback/src/test/java/com/apollographql/subscription/callback/SubscriptionCallbackHandlerTest.java

```diff
@@ -0,0 +1,369 @@
+            @Override
+            public MockResponse dispatch(@NotNull RecordedRequest recordedRequest) {
+              capturedRequests.add(recordedRequest.getBody().readUtf8());
+              return new MockResponse().setResponseCode(HttpStatus.NOT_FOUND.value());
+            }
+          });
+
+      var data =
+          Flux.just(1, 2)
+              .map((i) -> ExecutionResult.newExecutionResult().data(Map.of("counter", i)).build());
+      var handler = new SubscriptionCallbackHandler(new MockWebHandler(data));
+
+      var subscriptionId = UUID.randomUUID().toString();
+      var callbackUrl = server.url("/callback/" + subscriptionId).toString();
+      var verifier = "junit";
+      var callback = new SubscriptionCallback(callbackUrl, subscriptionId, verifier);
+
+      var graphQLRequest = stubWebGraphQLRequest(subscriptionId, callbackUrl);
+      var subscription = handler.handleSubscriptionUsingCallback(graphQLRequest, callback);
+      StepVerifier.create(subscription).expectNext(false).verifyComplete();
+
+      // wait for subscription to end
+      // TODO change to virtual time
+      Thread.sleep(500);
+
+      var objectMapper = new ObjectMapper();
+      Assertions.assertEquals(1, capturedRequests.size());
+      Assertions.assertEquals(
+          new CallbackMessageCheck(subscriptionId, callbackUrl),
+          objectMapper.readValue(capturedRequests.get(0), CallbackMessageCheck.class));
+    } catch (IOException | InterruptedException e) {
+      // failed to close the server
+    }
+  }
+
+  @Test
+  public void heartbeat() {
+    var capturedRequests = new ArrayList<String>();
+    try (var server = new MockWebServer()) {
+      server.setDispatcher(
+          new Dispatcher() {
+            @NotNull
+            @Override
+            public MockResponse dispatch(@NotNull RecordedRequest recordedRequest) {
+              capturedRequests.add(recordedRequest.getBody().readUtf8());
+              return new MockResponse()
+                  .setResponseCode(HttpStatus.OK.value())
+                  .setHeader(SUBSCRIPTION_PROTOCOL_HEADER, SUBSCRIPTION_PROTOCOL_HEADER_VALUE);
+            }
+          });
+
+      var subscriptionId = UUID.randomUUID().toString();
+      var callbackUrl = server.url("/callback/" + subscriptionId).toString();
+      var verifier = "junit";
+      var callback = new SubscriptionCallback(callbackUrl, subscriptionId, verifier);
+      var client = WebClient.builder().baseUrl(callback.callback_url()).build();
+
+      var graphQLRequest = stubWebGraphQLRequest(subscriptionId, callbackUrl);
+      var data =
+          Flux.just(1, 2)
+              .delayElements(Duration.ofMillis(3000))
+              .map((i) -> ExecutionResult.newExecutionResult().data(Map.of("counter", i)).build());
+      var handler = new SubscriptionCallbackHandler(new MockWebHandler(data));
+
+      // note: heartbeat goes into infinite recursion and does not emit value
+      // TODO update to use virtual timer
```

Comment:
unsure why but virtual timer didn't work... this can be fixed in the future (as it is right now those tests take couple seconds to run....)